### PR TITLE
symbols: support group references in replacements

### LINKED ISSUE #11107: Adding support for group matches in complexSymbols?
Hello,

In e.g. `source/locale/fr/symbols.dic`, we have 

```
dates .	((?<=\b\d\d)\.(?=\d\d.(\d{2}|\d{4})\b))|((?<=\b\d\d.\d\d)\.(?=(\d{2}|\d{4})\b))
dates .	 	all	always	# point de date
```

so as to remove dots from the dates. This is relatively complex, and could be much better written with matching groups:

```
dates .	\b(\d\d)\.(\d\d)\.(\d{2}|\d{4})\b
dates .	\1 \2 \3 	all	always	# point de date
```

It would allow to write more easily such rules like:

```
price	\b(\d\+),(\d\+)\s?€\b
price	\1 euros \2	all	always	# prix à virgules
```

etc.

I would be just a few lines more in _regexpRepl to perform the group replacement, I can work on it if the principle is Ok? (basically it means making the replacement string interpret `\` characters, thus allowing for this feature, and possibly even more in the future)

Samuel

### PR DESCRIPTION
**Pull request by sthibaul**
### Link to issue number:

fixes #11107

### Summary of the issue:

Rules such as `dates .` are quite complex to write: `((?<=\b\d\d)\.(?=\d\d.(\d{2}|\d{4})\b))|((?<=\b\d\d.\d\d)\.(?=(\d{2}|\d{4})\b))` because one wants to change two pieces of text here, and so we have to match them twice. Other rules are even more complex to write (e.g. `1,20€` -> `1 euro 20` needs two complex symbols to perform the two differing `,`->` euro ` `€`->`` replacements.

### Description of how this pull request fixes the issue:

This adds support for regexp group references.

### Testing performed:

I tested the use of it in the French translation, included in the commit:

```
dates .        \b(\d\d)\.(\d\d)\.(\d{2}|\d{4})\b
dates /        \b(\d\d)/(\d\d)/(\d{2}|\d{4})\b
```
```
dates .        \1 point \2 point \3        all     always  # point de date
dates /        \1 barre oblique \2 barre oblique \3        all     always  # barre oblique de date
```

The change uses Python's standard support for group, already used for the complex group parsing.

### Known issues with pull request:

It requires checking that no symbols file includes a backslash in the replacement string. I did this, and found two (commented) occurrences in the italian symbols file, fixed by the commit. (and incidently found a typo in the hebrew file)

### Change log entry:

*New features*
`Symbol pronunciation: Support grouping in complex symbol definition and group references in replacement rule, making them more powerful and simpler. (#11107)`

### CODE REVIEW

**Comment by josephsl** [comment]
Hi, this is more towards changes for developers, although it would be beneficial to talk about it briefly as part of the user guide if warranted. Thanks.

**Comment by sthibaul** [comment]
> Hi, this is more towards changes for developers, 

Do you mean that I shouldn't have submitted the pull request here, but rather to another repository?

**Comment by josephsl** [comment]
Hi, no, my comment relates to changes entry. Thanks.

**Comment by sthibaul** [comment]
I added some words in the ./devDocs/developerGuide.t2t

**Comment by sthibaul** [comment]
> Hi, no, my comment relates to changes entry. 

Ah, ok, I reworded it

**Comment by sthibaul** [comment]
> Hi, no, my comment relates to changes entry.

Maybe https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples#change-log-entry should say it explicitly that the text should be meant for users?

**Comment by Andre9642** [comment]
Hi Samuel :)
It seems to me that modifications made in 'locale/{fr,he,it}/symbols.dic' should be done via Assembla exclusively.
Would you like to join the team of translators?

**Comment by sthibaul** [comment]
Heya André :)
I can drop these from the MR, none of them actually have any effect, it was more for documentation.
I don't really plan to join the translators team, I just meant to pass on at least the date example so people get the idea and can then take benefit from it.

**Comment by feerrenrut** [comment]
@josephsl can you comment on whether it will cause problems for the current translation system if changes are accepted to languages other than English?

**Comment by josephsl** [comment]
Hi, as translations come from SVN and are done by individual translators, it would be best to push to English symbols file, which will then pass onto translators as a diff later. Thanks,

**Comment by sthibaul** [comment]
Here there are no english changes, just fixes in a couple of translations, and an improvement in another one.

**Comment by sthibaul** [comment]
To make things simpler, I have now dropped the locale/ changes, and we can commit the locale/ fixes separately

**Comment by Adriani90** [comment]
Will this also fix issue #2676?

**Comment by sthibaul** [comment]
> Will this also fix issue #2676?

No, it's unrelated, the grouping mentioned here is not of the same kind as in #2676

**Comment by sthibaul** [comment]
So, any comment on including this?

I would like to include the same support in speech-dispatcher which uses the same dict files, but I would rather avoid making the file format diverge...

**Comment by sthibaul** [comment]
Hello, any comment on this?

File: source/characterProcessing.py
Comment (feerrenrut) [inline source/characterProcessing.py:564]: Please extract this block to a named function. Did you look to see if the regex module can handle these escapes directly? I'm concerned this approach will be error prone. If this must be done manually, please add unit tests.

**Review by feerrenrut — CHANGES_REQUESTED** [review]
Thanks for the contribution @sthibaul and apologies for the delay to review the work.

File: devDocs/developerGuide.t2t
Comment (sthibaul) [reply devDocs/developerGuide.t2t]: Right, done so.

File: source/characterProcessing.py
Comment (sthibaul) [reply source/characterProcessing.py:553]: Unfortunately `\` seems to be still interpreted in raw strings:

```
>>> r'\'
  File "<stdin>", line 1
    r'\'
       ^
SyntaxError: EOL while scanning string literal
```

File: source/characterProcessing.py
Comment (sthibaul) [reply source/characterProcessing.py:564]: Ah, let me check, it seems like m.expand could do it.

Comment (sthibaul) [reply source/characterProcessing.py:564]: No, it doesn't work, because the numbers of the references that are used by `m.expand` are relative to the whole regexp, and not relative to the regexp inside the named group.

Comment (sthibaul) [reply source/characterProcessing.py:564]: (i.e. it's the `m+lastindex + ` here which is missing in the `re`-provided implementation)

Comment (sthibaul) [reply source/characterProcessing.py:564]: Is the dic engine actually tested at all? I don't find anything existing in the repository.

Comment (sthibaul) [reply source/characterProcessing.py:564]: I have added a test, but I don't know why the python linter doesn't like the heading comments, which are just exactly like in other files.

(there is also a fix in the french locale file, I have contacted the translator to get it fixed ahead of this)

Comment (lukaszgo1) [reply source/characterProcessing.py]: Commends in other files were mostly introduced before we introduced the linter unfortunately. You need to insert space after each hash sign to make it happy and add one more blank line before 
`class TestComplex(unittest.TestCase):`

Also were no longer placing the file name as the first line in the copyright header so please remove it.

Comment (sthibaul) [reply source/characterProcessing.py]: Ah, the linter only looks at new code, that's why. Ok, pushed a fixed version, thanks for the explanation.

File: tests/unit/test_characterProcessing.py
Comment (feerrenrut) [inline tests/unit/test_characterProcessing.py:18]: This is a good start! Thanks for doing it. Apologies if you are already planning on expanding this, but I really want to also see tests for the `_replaceGroups` method. That's the part that contains the custom code. Please try to ensure each branch is hit with the tests.

File: source/characterProcessing.py
Comment (feerrenrut) [inline source/characterProcessing.py:528]: Is there a reason not to use a raw string here (to remove confusion around the number of slash characters required)? 
```suggestion
				if c == r'\':

```

Comment (sthibaul) [reply source/characterProcessing.py:528]: This was already raised in another comment: that doesn't work, \ is still interpreted in raw strings:

>>> r'\'
  File "<stdin>", line 1
    r'\'
       ^
SyntaxError: EOL while scanning string literal

File: tests/unit/test_characterProcessing.py
Comment (sthibaul) [reply tests/unit/test_characterProcessing.py:36]: The test_group function was already triggering a call to `_replaceGroups`. But that wasn't exercising the '\\' case in group replacement so I have added a second test.

File: source/characterProcessing.py
Comment (feerrenrut) [reply source/characterProcessing.py:528]: Ah. Apologies.

File: source/characterProcessing.py
Comment (feerrenrut) [inline source/characterProcessing.py:523]: Please add type hints and a doc string for this function. Generally we prefer more descriptive variable names.

File: tests/unit/test_characterProcessing.py
Comment (feerrenrut) [inline tests/unit/test_characterProcessing.py:22]: Why is this specific to date?

Comment (sthibaul) [reply tests/unit/test_characterProcessing.py:22]: It is just the name of the group below, I could have put `bar` there and put `(?P<bar>` instead of `(?P<date>`, or whatever foobar-like name you'd prefer.

File: source/characterProcessing.py
Comment (sthibaul) [reply source/characterProcessing.py]: Ok, done so

File: tests/unit/test_characterProcessing.py
Comment (feerrenrut) [inline tests/unit/test_characterProcessing.py:28]: While this may test most of the paths (which is quite hard to determine at a glance), and is a good "happy path" test. I would like to see smaller examples that shows each of the paths. There are at least 6 paths through the `_replaceGroups` function, and many more combinations due to the internal loop. Please add an independent test for each path it can take. 

These tests act as documentation for what the expectations are for the function, so an unfamiliar developer can easily understand the different things the function can do.

Please also test the error case.

File: source/characterProcessing.py
Comment (feerrenrut) [inline source/characterProcessing.py:522]: Rather than in the doc string, please add this as a type annotation.
```suggestion
	def _replaceGroups(self, m: re.Matchobject, string: str) -> str:

```

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: tests/unit/test_characterProcessing.py
Comment (feerrenrut) [inline tests/unit/test_characterProcessing.py:41]: Thanks for adding extra tests, at the moment it is hard to read the 'actual' and 'expected' values.There is a lot happening on these lines. Could you split them in the following way please?
```suggestion
		replaced = self._replace(
			string="1",
			pattern=r"(?P<foo>(\d))",  # name group 'foo' looking for a single digit.
			name="foo",
			replacement="a"
		)
		self.assertEqual(replaced, "a")

```

File: tests/unit/test_characterProcessing.py
Comment (feerrenrut) [inline tests/unit/test_characterProcessing.py:55]: Can you explain here why the error is raised (for other error cases too please)

File: tests/unit/test_characterProcessing.py
Comment (feerrenrut) [inline tests/unit/test_characterProcessing.py:63]: I think this would be clearer if it didn't use '2' and '1', since these match the number in the group. Perhaps something like:
```suggestion
		self.assertEqual(self._replace("foo.BAR", r"(?P<foo>([a-z])\.([A-Z))", "foo", r"\2>\1"), "BAR>foo")

```

File: source/characterProcessing.py
Comment (feerrenrut) [inline source/characterProcessing.py:524]: Is there a test for the `\\\\` case?

File: tests/unit/test_characterProcessing.py
Comment (sthibaul) [reply tests/unit/test_characterProcessing.py:55]: I don't understand what kind of explanation you would like. Something non-supported shall report an error, I don't see what else to explain? I guess I can say that an escape at the end of a string is meaningless since we don't know what is being escaped, is that really about it?

File: tests/unit/test_characterProcessing.py
Comment (sthibaul) [reply tests/unit/test_characterProcessing.py]: Ok, I have separated out the tests.

File: source/characterProcessing.py
Comment (sthibaul) [reply source/characterProcessing.py]: Ok, done so

File: tests/unit/test_characterProcessing.py
Comment (sthibaul) [reply tests/unit/test_characterProcessing.py]: Ok, done so

File: tests/unit/test_characterProcessing.py
Comment (sthibaul) [reply tests/unit/test_characterProcessing.py]: Ok, done so.

File: source/characterProcessing.py
Comment (sthibaul) [reply source/characterProcessing.py:524]: I added one indeed, better check that quadruple backslash works.
Note that here in the doc string, `\\\\` is actually only two characters.

File: tests/unit/test_characterProcessing.py
Comment (feerrenrut) [reply tests/unit/test_characterProcessing.py:83]: If the comment can answer "why?": Why is an escape at the end of the string an error?

Comment (feerrenrut) [reply tests/unit/test_characterProcessing.py:83]: Not only that, but the comment currently says "at the end of the string" but the string parameter has no escape. It would be clearer if it said replacement? The docs could show a slightly modified valid version.

File: tests/unit/test_characterProcessing.py
Comment (feerrenrut) [inline tests/unit/test_characterProcessing.py:99]: I've just noticed that all these tests use name, when it is not strictly necessary to test the replace function. Why is this? I want to see tests for the custom code in the `_replaceGroups` code, which doesn't consider group name at all. The more simple the regex, the easier these tests will be to understand.

File: tests/unit/test_characterProcessing.py
Comment (sthibaul) [reply tests/unit/test_characterProcessing.py]: Ok, fixed so.

File: tests/unit/test_characterProcessing.py
Comment (sthibaul) [reply tests/unit/test_characterProcessing.py]: Right, we can just put anonymous groups. The only exception if of course the test for multiple groups, we want to name the one we aim to replace.
Done so.

**Comment by feerrenrut** [comment]
Hi @sthibaul can I request that you don't squash and force push every change? It makes it difficult to see what has changed since I last reviewed. Rest assured that we will squash merge to master, or if you want to clean the history of the branch please do so after the reviews are all complete.

File: source/locale/fr/symbols.dic
Comment (feerrenrut) [inline source/locale/fr/symbols.dic:43]: @sthibaul Interesting to note that this dates replacement is not done in the English dictionary! 
@josephsl Will this modification cause a problem for translators?

**Comment by zstanecic** [comment]
Yes, it may cause problems.

**Comment by sthibaul** [comment]
> Hi @sthibaul can I request that you don't squash and force push every change? It makes it difficult 

Ah, ok, sure, I thought github was providing an easy way to see the changes since last push (iirc gitlab does)

File: source/locale/fr/symbols.dic
Comment (sthibaul) [reply source/locale/fr/symbols.dic:43]: As I mentioned previously, this is a bug in the french translation, I'm discussing with the translator, so we get that included before this gets commited (without having to do the change, then, since it'll already be there).

Comment (feerrenrut) [reply source/locale/fr/symbols.dic:43]: I see. I'll add the blocked label to this PR so that it isn't merged until the translation change comes through. Please add a comment when that happens.

**Comment by CyrilleB79** [comment]
Seems this PR is blocked due to French translation?

With NVDA 2020.2, reading "20/08/2020" with punctuation level set to all, we can hear the \1 \2 and \3 expressions. Maybe an unintentional modification in commit 56765 of assembla SVN translation repo.

@MichelSuch, since you are the author of this Assembla commit, would you mind to confirm my assumption and revert the incorrect date rule. Then inform here so that this PR may be unblocked. Thanks.

**Comment by sthibaul** [comment]
> and revert the incorrect date rule

Well, I wouldn't say revert. We have a kind of chicken-and-egg here: we want to have a test that checks that translation rules do work, so we need a translation that has a rule using groups. French happens to do have one, even if for now it is bogus since the support is not there. But we can just unbreak that instead of reverting.

The question was rather about the "always" vs "norep", we are discussing the matter by mail (in French)

**Comment by sthibaul** [comment]
But yes, you are right in pinging @MichelSuch , this PR is waiting for his feedback to my mails :)

**Comment by CyrilleB79** [comment]
I may not have understood the whole issue.
Anyway IMO we should not commit permanently test stuff that introduce a regression here.
You can always attach a modified test symbols.dic file to this PR if needed and describe the test procedure in this PR.

Feel free to add me in the French e-mail exchange to discuss further the French file issue. I also belong to French translation team, even if my contribution is very minimal with respect to Michel's or Rémy's ones.

**Comment by sthibaul** [comment]
> Anyway IMO we should not commit permanently test stuff that introduce a regression here.

It was not actually meant to be commited, that was due to a misunderstanding.

> You can always attach a modified test symbols.dic file to this PR if needed and describe the test procedure in this PR.

The test is automatic, that's the whole point: to make sure that the code behave as expected. But I'm precisely exchanging with Michel to determine what is actually expected, because it does not seem that clear.

> Feel free to add me in the French e-mail exchange to discuss further the French file issue.

I don't know your email address. You can contact me on samuel.thibault@ens-lyon.org

**Comment by sthibaul** [comment]
Ok, we have sorted out the french translation question. Michel has commited reverting to not using groups, and using norep instead of always.

In this pull request I have dropped the fr_FR test, so that it can  be commited without waiting for the french translation update. 

When this pull request is merged, Michel will be able to update the french translation to use the group support, and then I'll submit another pull request to add a test for it.

**Comment by CyrilleB79** [comment]
I did not review the code. However I have some remarks:
* At first glance, the provided example does not seem realistic: why would a user replace the dot by space when the punctuation level is set to all? I would have expected something to be announced instead at this high level. The replacement rule could be for example (in French):
`dates .	\1 point \2 point \3 	all	always	# point de date`
  Maybe I have missed something, so please add an indication to let the reader know the context, i.e. why this rule was added in the French symbol file. @MichelSuch?
* Also the semantic of this rule has changed:
  * Before the rule was matching a dot in the date and replacing it:
  `dates .	 point 	all	norep	# point de date`
  * Now it matches the whole date and replace it with group substitution:
  `dates .	\1 \2 \3 	all	norep	# point de date`
  Thus I suggest renaming the rule from "dates ." to "date with ." in the example as well as in the translation repo French file when this file will have to be modified (after this PR is merged). Also the associated comment should specify "date avec point" (i.e. "date with dot") instead of "point de date" (i.e. "date dot")
* Regarding change log entry something more precise would help such as:
`Symbol pronunciation: Support grouping in complex symbol definition and group references in replacement rule`

**Comment by sthibaul** [comment]
Any news on this?  @feerrenrut perhaps? The french translation question has been sorted out, we will update it after this gets commited.

**Comment by feerrenrut** [comment]
Ok, thanks @sthibaul so I can take it that this is no longer blocked by the French translation and is ready for review and merging?

**Review by feerrenrut — DISMISSED** [review]

**Review by feerrenrut — APPROVED** [review]

**Pull request merged**
